File: nvxapp.client/src/nvx/ClientServer-Service/GestionePresenze/Par_Competenza/par-competenza.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { AuthService } from '../../../Utility/infrastructure/auth.service';
import { environment } from '../../../../environments/environment';
import { GenericRequest } from '../../ModelsBase/generic-request';
import { GenericResult } from '../../ModelsBase/generic-result';
import { map, Observable } from 'rxjs';
import { Par_Competenza_GetAll_InModel, Par_Competenza_GetAll_OutModel, Par_CompetenzaPutInModel, Par_CompetenzaPutOutModel, Par_CompetenzaModel, Par_CompetenzaGetInModel, Par_CompetenzaGetOutModel, Par_CompetenzaDeleteInModel, Par_CompetenzaDeleteOutModel } from './Models/par-competenza-model';


@Injectable({
  providedIn: 'root'
})
export class ParCompetenzaService {
  constructor(private http: HttpClient,
              private authService: AuthService) { }

  GetAll(model: GenericRequest<Par_Competenza_GetAll_InModel>): Observable<GenericResult<Par_Competenza_GetAll_OutModel>> {
    return this.http.post<GenericResult<Par_Competenza_GetAll_OutModel>>(environment.remoteData.apiUri + 'Par_Competenza/GetAll', model)
      .pipe(map(r => r));
  }

  Par_CompetenzaGet(model: GenericRequest<Par_CompetenzaGetInModel>): Observable<GenericResult<Par_CompetenzaGetOutModel>> {
    return this.http.post<GenericResult<Par_CompetenzaGetOutModel>>(environment.remoteData.apiUri + 'Par_Competenza/Par_CompetenzaGet', model)
      .pipe(map(r => r));
  }

  Par_CompetenzaPut(model: GenericRequest<Par_CompetenzaPutInModel>): Observable<GenericResult<Par_CompetenzaPutOutModel>> {
    return this.http.post<GenericResult<Par_CompetenzaPutOutModel>>(environment.remoteData.apiUri + 'Par_Competenza/Par_CompetenzaPut', model)
      .pipe(map(r => r));
  }

  Par_CompetenzaDelete(model: GenericRequest<Par_CompetenzaDeleteInModel>): Observable<GenericResult<Par_CompetenzaDeleteOutModel>> {
    return this.http.post<GenericResult<Par_CompetenzaDeleteOutModel>>(environment.remoteData.apiUri + 'Par_Competenza/Par_CompetenzaDelete', model)
      .pipe(map(r => r));
  }


}

File: nvxapp.server.service/ClientServer-Service/GestionePresenze/Par_CompetenzaService/Par_CompetenzaService.cs
using AutoMapper;
using Microsoft.AspNetCore.Http;
using Microsoft.AspNetCore.Identity;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.Options;
using nvxapp.server.Base;
using nvxapp.server.data.Entities.Public;
using nvxapp.server.data.Entities.Tenant;
using nvxapp.server.data.Repositories.Public;
using nvxapp.server.data.Repositories.Tenant.GestionePresenze;
using nvxapp.server.service.ClientServer_Service.GestionePresenze._utility;
using nvxapp.server.service.ClientServer_Service.GestionePresenze.Par_CompetenzaService.Models;
using nvxapp.server.service.ClientServer_Service.ModelsBase;
using nvxapp.server.service.Interfaces;
using nvxapp.server.service.ServerModels;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace nvxapp.server.service.ClientServer_Service.GestionePresenze.Par_CompetenzaService
{
    public class Par_CompetenzaService : ServiceBase, IPar_CompetenzaService
    {
        private readonly IPar_CompetenzaRepository _par_CompetenzaRepository;
        private readonly IGestionePresenzeUserUtility _gestionePresenzeUserUtility;

        public Par_CompetenzaService(IMapper mapper,
                                  UserManager<ApplicationUser> userManager,
                                  IAspNetUsersRepository aspNetUsersRepository,
                                  IOptions<JwtParameter> jwtParameter,
                                  IHttpContextAccessor httpContextAccessor,
                                  IConfiguration configuration,
                                  IGestionePresenzeUserUtility gestionePresenzeUserUtility,
                                  IPar_CompetenzaRepository az_CompetenzaRepository) : base(mapper, userManager, aspNetUsersRepository, jwtParameter, configuration, httpContextAccessor)
        {
            _par_CompetenzaRepository = az_CompetenzaRepository;
            _gestionePresenzeUserUtility = gestionePresenzeUserUtility;
        }

        public virtual async Task<GenericResult<Par_Competenza_GetAll_OutModel>> GetAll(GenericRequest<Par_Competenza_GetAll_InModel> model, bool isSubProcess)
        {
            return await ExecuteAction(model, async () =>
            {
                Par_Competenza_GetAll_OutModel retVal = new Par_Competenza_GetAll_OutModel();

                int IdCompany;
                int.TryParse(this.CurrentCompany, out IdCompany);

                Company_DATA_COMB_AzAna_AzSedi_AzReparto_Az_Cfg company_DATA = await _gestionePresenzeUserUtility.Get_AzAna_AzSedi_AzReparto_Az_Cfg(IdCompany, true);
                if (company_DATA != null && company_DATA.az_Anagrafica != null)
                {
                    var az_Competenza = _par_CompetenzaRepository.FindAll(x => x.IdAz_Anagrafica == company_DATA.az_Anagrafica.Id).ToList();
                    retVal.Par_Competenza = _mapper.Map<List<Par_CompetenzaModel>>(az_Competenza);
                }

                await Task.Delay(DelayAsyncMethod);
                return retVal;
            }, isSubProcess);
        }

        public virtual async Task<GenericResult<Par_CompetenzaGetOutModel>> Par_CompetenzaGet(GenericRequest<Par_CompetenzaGetInModel> model, bool isSubProcess)
        {
            return await ExecuteAction(model, async () =>
            {
                Par_CompetenzaGetOutModel retVal = new Par_CompetenzaGetOutModel();
                var az_Competenza = await _par_CompetenzaRepository.FindByIdAsync(model.Data.Id);
                if (az_Competenza != null)
                {
                    retVal.Par_Competenza = _mapper.Map<Par_CompetenzaModel>(az_Competenza);
                }
                else
                {
                    retVal.Par_Competenza = new Par_CompetenzaModel();
                }
                await Task.Delay(DelayAsyncMethod);
                return retVal;
            }, isSubProcess);
        }

        public virtual async Task<GenericResult<Par_CompetenzaPutOutModel>> Par_CompetenzaPut(GenericRequest<Par_CompetenzaPutInModel> model, bool isSubProcess)
        {
            return await ExecuteAction(model, async () =>
            {
                Par_CompetenzaPutOutModel retVal = new Par_CompetenzaPutOutModel();
                retVal.Par_Competenza = model.Data.Par_Competenza;

                int IdCompany;
                int.TryParse(this.CurrentCompany, out IdCompany);
                Company_DATA_COMB_AzAna_AzSedi_AzReparto_Az_Cfg company_DATA = await _gestionePresenzeUserUtility.Get_AzAna_AzSedi_AzReparto_Az_Cfg(IdCompany, true);
                if (company_DATA != null && company_DATA.az_Anagrafica != null)
                {
                    var az_Competenza = await _par_CompetenzaRepository.FindByIdAsync(model.Data.Par_Competenza.Id);
                    if (az_Competenza == null)
                    {
                        az_Competenza = _mapper.Map<Par_Competenza>(model.Data.Par_Competenza);
                        az_Competenza.IdAz_Anagrafica = company_DATA.az_Anagrafica.Id;
                    }
                    else
                    {
                        az_Competenza = _mapper.Map<Par_Competenza>(model.Data.Par_Competenza);
                    }
                    az_Competenza = await _par_CompetenzaRepository.UpsertAsync(az_Competenza);
                    retVal.Par_Competenza = _mapper.Map<Par_CompetenzaModel>(az_Competenza);
                }
                await Task.Delay(DelayAsyncMethod);
                return retVal;
            }, isSubProcess);
        }
    }

    public interface IPar_CompetenzaService : IServiceBase
    {
        Task<GenericResult<Par_Competenza_GetAll_OutModel>> GetAll(GenericRequest<Par_Competenza_GetAll_InModel> model, bool isSubProcess);
        Task<GenericResult<Par_CompetenzaGetOutModel>> Par_CompetenzaGet(GenericRequest<Par_CompetenzaGetInModel> model, bool isSubProcess);
        Task<GenericResult<Par_CompetenzaPutOutModel>> Par_CompetenzaPut(GenericRequest<Par_CompetenzaPutInModel> model, bool isSubProcess);
    }
}
